src/simulations/optimisation.py

```diff
@@ -150,24 +175,84 @@ def _power_flow_equations(self, p, netp, grid, hourly_line_losses_pu):
             'undervoltage_costs', (self.grd['n_buses'] - 1, self.N), vtype='continuous'
         )
 
-        # active and reactive loads: netp and netq from kW to W (*1000) to per unit system (/Ab)
-        p.add_list_of_constraints(
-            [
-                pi[:, t] == self.grd['flex_buses'] * netp[:, t] * 1000 / self.grd['base_power']
-                for t in range(self.N)
-            ]
-        )
-        p.add_list_of_constraints(
-            qi[:, t] == self.grd['flex_buses'] * netq[:, t] * 1000 / self.grd['base_power']
-            for t in range(self.N)
-        )
+        # active and reactive loads
+        # passive homes: heat, home and car
+        if self.n_homesP > 0:
+            p.add_constraint(
+                [q_heat_home_car_non_flex == self._calculate_reactive_power(
+                    self.loads['netp0'],
+                    self.grd['pf_non_flex_heat_home_car'])
+                    ])
+        # flex houses: heat and home
+        p.add_constraint(q_heat_home_flex == self._calculate_reactive_power(
+            totcons, self.grd['pf_flexible_heat_home']))
+
+        # flex houses: car
+        p.add_constraint(p_car_flex == charge / self.car['eta_ch'] + discharge_other)
+        p.add_constraint(p_car_flex <= self.grd['max_active_power_car'])
+
+        # if we don't allow the use of the battery reactive power for control
+        # then we restain it by using the power factor
+        if self.reactive_power_for_voltage_control:
+            for time in range(self.N):
+                p.add_list_of_constraints([
+                    p_car_flex2[home, time] >= p_car_flex[home, time]
+                    * p_car_flex[home, time] for home in range(self.n_homes)
+                ])
+                p.add_list_of_constraints([
+                    q_car_flex2[home, time] >= q_car_flex[home, time]
+                    * q_car_flex[home, time] for home in range(self.n_homes)
+                ])
+                p.add_list_of_constraints([
+                    p_car_flex2[home, time] + p_car_flex2[home, time]
+                    <= self.grd['max_apparent_power_car']**2 for home in range(self.n_homes)
+                ])
+        else:
+            p.add_constraint(q_car_flex == self._calculate_reactive_power(
```

Comment:
does it work to have a function in a constraint? I would have assumed not? but never tried